File: Amplitude Hackathon Summer/2024/P4.cpp
#include <iostream>
#include <vector>
#include <algorithm>

bool can_fit(int width, int n, int l, int s, const std::vector<int>& word_widths) {
    int lines = 1;
    int current_line_width = 0;
    int words_in_line = 0;

    for (int w : word_widths) {
        if (words_in_line > 0) {
            if (current_line_width + s + w <= width) {
                current_line_width += s + w;
                words_in_line++;
            } else {
                lines++;
                if (lines > l) {
                    return false;
                }
                current_line_width = w;
                words_in_line = 1;
            }
        } else {
            current_line_width = w;
            words_in_line = 1;
        }
    }

    return true;
}

int main() {
    int n, l, s;
    std::cin >> n >> l >> s;
    std::vector<int> word_widths(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> word_widths[i];
    }

    int left = *std::max_element(word_widths.begin(), word_widths.end());
    int right = std::accumulate(word_widths.begin(), word_widths.end(), 0) + (n - 1) * s;
    
    while (left < right) {
        int mid = left + (right - left) / 2;
        if (can_fit(mid, n, l, s, word_widths)) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }

    std::cout << left << std::endl;

    return 0;
}

File: Amplitude Hackathon Summer/2024/P4.py
# Amplitude Hackathon Summer '24 Problem 4 - Spenser's Big Announcement
# DMOJ: https://dmoj.ca/problem/ampl2024sp4

def can_fit(width, n, l, s, word_widths):
    lines = 1
    current_line_width = 0
    words_in_line = 0

    for w in word_widths:
        if words_in_line > 0:
            if current_line_width + s + w <= width:
                current_line_width += s + w
                words_in_line += 1
            else:
                lines += 1
                if lines > l:
                    return False
                current_line_width = w
                words_in_line = 1
        else:
            current_line_width = w
            words_in_line = 1

    return True


n, l, s = map(int, input().split())
word_widths = list(map(int, input().split()))

left = max(word_widths)
right = sum(word_widths) + (n - 1) * s
while left < right:
    mid = (left + right) // 2
    if can_fit(mid, n, l, s, word_widths):
        right = mid
    else:
        left = mid + 1


print(left)